File: python/fedml/arguments.py
# Copyright 2022, FedML.  All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Arguments."""

import argparse
import os
from os import path
import logging

import yaml

from .constants import (
    FEDML_TRAINING_PLATFORM_SIMULATION,
    FEDML_SIMULATION_TYPE_MPI,
    FEDML_SIMULATION_TYPE_SP,
    FEDML_TRAINING_PLATFORM_CROSS_SILO,
    FEDML_TRAINING_PLATFORM_CROSS_DEVICE,
    FEDML_CROSS_SILO_SCENARIO_HIERARCHICAL,
    FEDML_TRAINING_PLATFORM_CHEETAH,
    FEDML_TRAINING_PLATFORM_SERVING,
)


def add_args():
    parser = argparse.ArgumentParser(description="FedML")
    parser.add_argument(
        "--yaml_config_file",
        "--cf",
        help="yaml configuration file",
        type=str,
        default="",
    )

    # default arguments
    parser.add_argument("--run_id", type=str, default="0")

    # default arguments
    parser.add_argument("--rank", type=int, default=0)

    # default arguments
    parser.add_argument("--local_rank", type=int, default=0)
    
    # For hierarchical scenario
    parser.add_argument("--node_rank", type=int, default=0)

    # default arguments
    parser.add_argument("--role", type=str, default="client")

    args, unknown = parser.parse_known_args()
    return args


class Arguments:
    """Argument class which contains all arguments from yaml config and constructs additional arguments"""

    def __init__(self, cmd_args, training_type=None, comm_backend=None, override_cmd_args=True):
        # set the command line arguments
        cmd_args_dict = cmd_args.__dict__
        for arg_key, arg_val in cmd_args_dict.items():
            setattr(self, arg_key, arg_val)

        self.get_default_yaml_config(cmd_args, training_type, comm_backend)
        if not override_cmd_args:
            # reload cmd args again
            for arg_key, arg_val in cmd_args_dict.items():
                setattr(self, arg_key, arg_val)
    def load_yaml_config(self, yaml_path):
        with open(yaml_path, "r") as stream:
            try:
                return yaml.safe_load(stream)
            except yaml.YAMLError as exc:
                raise ValueError("Yaml error - check yaml file")

    def get_default_yaml_config(self, cmd_args, training_type=None, comm_backend=None):
        if cmd_args.yaml_config_file == "":
            path_current_file = path.abspath(path.dirname(__file__))
            if (
                training_type == FEDML_TRAINING_PLATFORM_SIMULATION
                and comm_backend == FEDML_SIMULATION_TYPE_SP
            ):
                config_file = path.join(
                    path_current_file, "config/simulation_sp/fedml_config.yaml"
                )
                cmd_args.yaml_config_file = config_file
                print(
                    "training_type == FEDML_TRAINING_PLATFORM_SIMULATION and comm_backend == FEDML_SIMULATION_TYPE_SP"
                )
            elif (
                training_type == FEDML_TRAINING_PLATFORM_SIMULATION
                and comm_backend == FEDML_SIMULATION_TYPE_MPI
            ):
                config_file = path.join(
                    path_current_file, "config/simulaton_mpi/fedml_config.yaml"
                )
                cmd_args.yaml_config_file = config_file
                print(
                    "training_type == FEDML_TRAINING_PLATFORM_SIMULATION and comm_backend == FEDML_SIMULATION_TYPE_MPI"
                )
            elif training_type == FEDML_TRAINING_PLATFORM_CROSS_SILO:
                print("training_type == FEDML_TRAINING_PLATFORM_CROSS_SILO")
            elif training_type == FEDML_TRAINING_PLATFORM_CROSS_DEVICE:
                print("training_type == FEDML_TRAINING_PLATFORM_CROSS_DEVICE")
            elif training_type == FEDML_TRAINING_PLATFORM_CHEETAH:
                print("training_type == FEDML_TRAINING_PLATFORM_CHEETAH")
            elif training_type == FEDML_TRAINING_PLATFORM_SERVING:
                print("training_type == FEDML_TRAINING_PLATFORM_SERVING")
            else:
                raise Exception(
                    "no such a platform. training_type = {}, backend = {}".format(
                        training_type, comm_backend
                    )
                )

        self.yaml_paths = [cmd_args.yaml_config_file]
        # Load all arguments from yaml config
        # https://www.cloudbees.com/blog/yaml-tutorial-everything-you-need-get-started
        configuration = self.load_yaml_config(cmd_args.yaml_config_file)

        # Override class attributes from current yaml config
        self.set_attr_from_config(configuration)

        if cmd_args.yaml_config_file == "":
            path_current_file = path.abspath(path.dirname(__file__))
            if (
                training_type == FEDML_TRAINING_PLATFORM_SIMULATION
                and comm_backend == FEDML_SIMULATION_TYPE_SP
            ):
                pass
            elif (
                training_type == FEDML_TRAINING_PLATFORM_SIMULATION
                and comm_backend == FEDML_SIMULATION_TYPE_MPI
            ):
                self.gpu_mapping_file = path.join(
                    path_current_file, "config/simulaton_mpi/gpu_mapping.yaml"
                )
            elif training_type == FEDML_TRAINING_PLATFORM_CROSS_SILO:
                pass
            elif training_type == FEDML_TRAINING_PLATFORM_CROSS_DEVICE:
                pass
            elif training_type == FEDML_TRAINING_PLATFORM_CHEETAH:
                pass
            elif training_type == FEDML_TRAINING_PLATFORM_SERVING:
                pass
            else:
                pass
        

        if hasattr(self, "training_type"):
            training_type = self.training_type

        if training_type == FEDML_TRAINING_PLATFORM_CROSS_SILO:
            if (
                hasattr(self, "scenario")
                and self.scenario == FEDML_CROSS_SILO_SCENARIO_HIERARCHICAL
            ):
                # Add extra configs specific to silos or server
                self.rank = int(self.rank)
                if self.rank == 0:
                    extra_config_path = self.server_config_path
                else:
                    extra_config_path = self.client_silo_config_paths[self.rank - 1]
                extra_config = self.load_yaml_config(extra_config_path)
                self.set_attr_from_config(extra_config)

        return configuration

    def set_attr_from_config(self, configuration):
        for _, param_family in configuration.items():
            for key, val in param_family.items():
                setattr(self, key, val)


def load_arguments(training_type=None, comm_backend=None):
    cmd_args = add_args()
    # Load all arguments from YAML config file
    args = Arguments(cmd_args, training_type, comm_backend)

    if not hasattr(args, "worker_num"):
        args.worker_num = args.client_num_per_round
        
    # os.path.expanduser() method in Python is used
    # to expand an initial path component ~( tilde symbol)
    # or ~user in the given path to user’s home directory.
    if hasattr(args, "data_cache_dir"):
        args.data_cache_dir = os.path.expanduser(args.data_cache_dir)
    if hasattr(args, "data_file_path"):
        args.data_file_path = os.path.expanduser(args.data_file_path)
    if hasattr(args, "partition_file_path"):
        args.partition_file_path = os.path.expanduser(args.partition_file_path)
    if hasattr(args, "part_file"):
        args.part_file = os.path.expanduser(args.part_file)

    args.rank = int(args.rank)
    return args

File: javascript/lib/fedml/arguments.js
// import { homedir } from 'os';
// import fs from 'fs';
const runArgsJSON = {
  configName: 'newPackage',
  packages_config: {
    linuxClient: 'client-package.zip',
    server: 'server-package.zip',
    androidClientVersion: 0,
    androidClientUrl: '',
    serverUrl: 'https://fedml.s3.us-west-1.amazonaws.com/1658063399409server-package.zip',
    androidClient: '',
    linuxClientUrl: 'https://fedml.s3.us-west-1.amazonaws.com/1658063403141client-package.zip',
  },
  data_config: {
    privateLocalData: '',
    syntheticDataUrl: '',
    syntheticData: '',
  },
  userId: 'fedml-yide',
  parameters: {
    model_args: {
      model_file_cache_folder: './model_file_cache',
      model: 'lr',
      global_model_file_path: './model_file_cache/global_model.pt',
    },
    device_args: {
      using_gpu: false,
      gpu_mapping_key: 'mapping_default',
      gpu_mapping_file: 'config/gpu_mapping.yaml',
    },
    comm_args: {
      grpc_ipconfig_path: './config/grpc_ipconfig.csv',
      backend: 'MQTT_S3',
    },
    train_args: {
      batch_size: 10,
      weight_decay: 0.001,
      client_num_per_round: 2,
      client_num_in_total: 2,
      comm_round: 50,
      client_optimizer: 'sgd',
      epochs: 1,
      learning_rate: 0.03,
      federated_optimizer: 'FedAvg',
    },
    environment_args: { bootstrap: 'config/bootstrap.sh' },
    validation_args: { frequency_of_the_test: 1 },
    common_args: {
      random_seed: 0,
      scenario: 'horizontal',
      training_type: 'cross_silo',
      config_version: 'release',
      using_mlops: false,
    },
    data_args: {
      partition_method: 'hetero',
      partition_alpha: 0.5,
      dataset: 'mnist',
      data_cache_dir: '~/fedml_data',
    },
    tracking_args: {
      wandb_project: 'fedml',
      wandb_name: 'fedml_torch_fedavg_mnist_lr',
      wandb_key: 'ee0b5f53d949c84cee7decbe7a629e63fb2f8408',
      enable_wandb: false,
      log_file_dir: './log',
    },
  },
}
export class Arguments {
  constructor(cmd_args, training_type = null, comm_backend = null, override_cmd_args = true) {
    // Object.keys(cmd_args).map((key) => {
    //   this[key] = cmd_args.key;
    // });
    for (const i in cmd_args)
      this[i] = cmd_args[i]
    this.get_default_yaml_config(cmd_args, training_type, comm_backend)
    if (!override_cmd_args) {
      for (const j in cmd_args)
        this[j] = cmd_args[j]
    }
  }

  async load_yaml_config() {
    try {
      const yaml_config = '{"configName":"newPackage","packages_config":{"linuxClient":"client-package.zip","server":"server-package.zip","androidClientVersion":0,"androidClientUrl":"","serverUrl":"https://fedml.s3.us-west-1.amazonaws.com/1658063399409server-package.zip","androidClient":"","linuxClientUrl":"https://fedml.s3.us-west-1.amazonaws.com/1658063403141client-package.zip"},"data_config":{"privateLocalData":"","syntheticDataUrl":"","syntheticData":""},"userId":"fedml-yide","parameters":{"model_args":{"model_file_cache_folder":"./model_file_cache","model":"lr","global_model_file_path":"./model_file_cache/global_model.pt"},"device_args":{"using_gpu":false,"gpu_mapping_key":"mapping_default","gpu_mapping_file":"config/gpu_mapping.yaml"},"comm_args":{"grpc_ipconfig_path":"./config/grpc_ipconfig.csv","backend":"MQTT_S3"},"train_args":{"batch_size":10,"weight_decay":0.001,"client_num_per_round":2,"client_num_in_total":2,"comm_round":50,"client_optimizer":"sgd","epochs":1,"learning_rate":0.03,"federated_optimizer":"FedAvg"},"environment_args":{"bootstrap":"config/bootstrap.sh"},"validation_args":{"frequency_of_the_test":1},"common_args":{"random_seed":0,"scenario":"horizontal","training_type":"cross_silo","config_version":"release","using_mlops":false},"data_args":{"partition_method":"hetero","partition_alpha":0.5,"dataset":"mnist","data_cache_dir":"~/fedml_data"},"tracking_args":{"wandb_project":"fedml","wandb_name":"fedml_torch_fedavg_mnist_lr","wandb_key":"ee0b5f53d949c84cee7decbe7a629e63fb2f8408","enable_wandb":false,"log_file_dir":"./log"}}}'
      const yamlObj = JSON.parse(yaml_config)
      return yamlObj
    }
    catch (error) {
      console.log(error)
    }
  }

  async get_default_yaml_config(cmd_args, training_type = null, comm_backend = null) {
    if (cmd_args.yaml_config_file == '') {
      if (training_type == 'cross_silo')
        console.log('training_type == FEDML_TRAINING_PLATFORM_CROSS_SILO')
    }
    const configuration = await this.load_yaml_config()
    this.set_attr_from_config(configuration)
    return configuration
  }

  set_attr_from_config(configuration) {
    for (const i in configuration) {
      if (typeof configuration[i] == 'object') {
        for (const j in configuration[i])
          this[j] = configuration[i][j]
      }
    }
  }

  static add_args() {
    const args = {
      yaml_config_file: 'config/fedml_config.yaml',
      run_id: 0,
      rank: 0,
      local_rank: 0,
      node_rank: 0,
      role: 'client',
    }
    return args
  }

  static load_arguments(training_type = null, comm_backend = null) {
    const cmd_args = this.add_args()
    const args = new Arguments(cmd_args, training_type, comm_backend)
    if (args.worker_num == undefined) {
      args.worker_num = 1
      args.client_num_per_round = 1
    }
    // if (args.data_cache_dir != undefined) {
    //   args.data_cache_dir = homedir.path.expanduser(args.data_cache_dir);
    // }
    // if (args.data_file_path != undefined) {
    //   args.data_file_path = homedir.path.expanduser(args.data_file_path);
    // }
    // if (args.partition_file_path != undefined) {
    //   args.partition_file_path = homedir.path.expanduser(args.partition_file_path);
    // }
    // if (args.part_file != undefined) {
    //   args.part_file = homedir.path.expanduser(args.part_file);
    // }
    args.rank = parseInt(args.rank)
    return args
  }
}
